tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+    # associated with two aten C++ functions: the base version and the
+    # out-place variant. Usually the two functions have the same set of
+    # arguments - of course, except for the output arguments. But in some
+    # cases they might have slightly different C++ argument types -
+    # affected by the 'use_c10_dispatcher' state.
+    #
+    # More specially, 'Tensor?' type can be translated into
+    # either 'const c10::optional<Tensor>&' or 'const Tensor &'.
+    # Unfortunately, this difference can affect how we should access arg
+    # parser output. The former expects '_r.optionalTensor(i)' while the
+    # latter expects '_r.tensor(i)'.
+    #
+    # Because of this subtle difference, we cannot solely use the shared
+    # python signature to determine the RHS expr for both C++ variants.
+    # We could create and use each C++ variant's own python signature,
+    # but we have to fix the argument index difference between the two
+    # python signatures like the old codegen does - and it feels wrong as
+    # technically there is only one shared python signature!
+    #
+    # So here we pass in the lambda wrapper's argument and use it to
+    # decide what PythonArgParser unpack method to use.
+    #
+    # TODO: this seems too complicated - maybe we can simplify after full
+    # c10 dispatch migration?
+    typename = la.cpp_type_str \
+        if a.name != 'out' and la is not None else a.cpp_type_str
+
+    if a.default_init is not None:
+        # Note: only introduced in tensor_options_args
+        if typename not in UNPACK_WITH_DEFAULT_METHODS:
+            raise RuntimeError(
+                f'type \'{typename}\' is not supported in default_init')
+        unpack_with_default = UNPACK_WITH_DEFAULT_METHODS[typename]
+        expr = f'_r.{unpack_with_default}({arg_index}, {a.default_init})'
+    elif a.size is not None:
+        if typename not in UNPACK_WITH_SIZE_METHODS:
+            raise RuntimeError(
+                f'type \'{typename}\' with definite size ({a.size}) is not supported')
+        unpack_with_size = UNPACK_WITH_SIZE_METHODS[typename].format(a.size)
+        expr = f'_r.{unpack_with_size}({arg_index})'
+    else:
+        unpack = UNPACK_METHODS.get(typename)
+        if unpack is None:
+            raise RuntimeError(f'type \'{typename}\' is not supported')
+        expr = f'_r.{unpack}({arg_index})'
+
+    return PythonArgParserOutputExpr(
+        name=a.name,
+        expr=expr,
+        index=arg_index,
+        argument=a,
+    )
+
+
+# argument name to 'simple_type' for scattered tensor options fields
+TENSOR_OPTIONS_FIELDS = {
+    'dtype': 'ScalarType',
+    'device': 'Device',
+    'layout': 'Layout',
+    'pin_memory': 'bool',
+    'requires_grad': 'bool',
+}
+
+
+def bind_python_cpp(
+    ps: PythonSignature, f: NativeFunction, method: bool
+) -> Tuple[
+    LocalInitExprs,
+    DispatchLambdaArgumentExprs,
+    Dict[str, PythonArgParserOutputExpr],
```

Comment:
Wondering what the str in this map denotes

Reply:
'str' is the name of the argument - it's for the old codegen to get the RHS expr of certain scattered arguments, e.g.: 'requires_grad', 'out'.

As you pointed out, this method is on the boundary of code/new codegen - I'll probably move some part of it to the old codegen to make the interface cleaner.

Reply:
rewrote this part